### ISSUE TITLE
Remove redirects in update actions

### ISSUE THREAD

**Issue by atsmith813** [forem/forem#10338]
## What is the problem?
Currently, we have a race condition any time we're using a redirect after an action that updates data causing the cache to be busted and we're directing to the page whose cache was just busted. We usually lose the race - we redirect before the cache can be successfully busted. This results in users seeing stale content.

The best example is updating a comment. Once a comment is updated, we bust the cache for the comment and redirect back to the comments Index action. The redirect happens before the busting of the cache causing the user to not see their new update unless they then refresh the page again manually.

There are a handful of places where this race condition poses an issue. It's mostly in update actions where we update cached data and then redirect to the index view where we're displaying said data right after.
- [x] [Updating a comment](https://github.com/forem/forem/blob/6c3f5d51fe20413682feea411bc40e522c97d90b/app/controllers/comments_controller.rb#L142-L144)
- [x] ~[Tag updates (only admins and tag mods can do this)](https://github.com/forem/forem/blob/6c3f5d51fe20413682feea411bc40e522c97d90b/app/controllers/tags_controller.rb#L22-L24)~
- [ ] [Updating articles under certain conditions](https://github.com/forem/forem/blob/6c3f5d51fe20413682feea411bc40e522c97d90b/app/controllers/articles_controller.rb#L151-L158)
- [ ] [Updating listings](https://github.com/forem/forem/blob/6c3f5d51fe20413682feea411bc40e522c97d90b/app/controllers/concerns/listings_toolkit.rb#L148-L149)
- [ ] Document this quirk

## Solution
To fix this, we'll refactor the code to not redirect.

**Comment by atsmith813** [comment]
To our awesome contributors, please note that this issue will be worked on by a Forem team member and is not up for grabs.

**Comment by citizen428** [comment]
> To fix this, we'll refactor the code to not redirect.

This means changing a pretty common Rails idiom. Pretty much every Rails tutorial on the planet will do the following in their update action:

1. Update the actual object
2. Set a flash message
3. Redirect to the original page

Two prominent examples:

* [The Rails Guides](https://edgeguides.rubyonrails.org/getting_started.html):
    ```ruby
    def update
      @article = Article.find(params[:id])

     if @article.update(article_params)
        redirect_to @article
      else
        render 'edit'
      end
    end
    ```
* [Michael Hartl's Rails tutorial](https://3rd-edition.railstutorial.org/book/updating_and_deleting_users), the most used/recommended Rails tutorial out there:
    ```ruby
     if @user.update_attributes(user_params)
       flash[:success] = "Profile updated"
        redirect_to @user
     else
     ```

Of course we don't have to blindly follow Rails conventions (there are places where we don't), but this seems a departure from something pretty fundamental that could throw off external contributors. Have we considered other options like bustin the cache inline before redirecting or adding a way to bypass the cache after an update?

**Comment by rhymes** [comment]
A couple of general question:

How slow is the inline cache busting?
Do we actually need the cache in those endpoints? Would it that bad if the page the user lands on is not cached? Obviously case by case.
Is there anything we can cache server side and stop using edge in there?

I understand the logic of this issue somewhat, I'm mostly wondering if the logical choice is a consequence of something we actually need or we can do away with.

**Comment by atsmith813** [comment]
@rhymes @citizen428 yep! I'm open to alternatives!

The TLDR is
> adding a way to bypass the cache after an update

is an ideal way to go, but that's proving far more difficult than we thought to do right now with our Openresty setup. I think we can get there at some point. The drawback to that approach that we considered is that it feels weird to see redirects in the server-side code, but the browser never sees it. It's a layer of complexity for our future selves and contributors to remember - there's sorcery at the edge which is a box that isn't visible.

> How slow is the inline cache busting?

It's less about speed and more about code maintainability. 2 things jump out to me about going inline.
1. If you turn into a synchronous blocking action and bring cache-busting inline, what do we do if/when cache-busting fails? We're back in the same boat. We can't redirect because we'd hit this same issue - we'd be serving stale content from the cache. The question then becomes how often might that happen, the cache failing to bust? It happens now with Fastly - it's pretty rare. I'd expect it to happen semi-often at first as we get our footing with Openresty, and then slow down to rare over time.
2. To @citizen428's 2 points
> pretty common Rails idiom

>  It also means everyone one the team would need to understand our caching, to be able to make the decision whether 
or not their update actions can be standard or need to follow the special case (in an ideal world that would be the case, I'm not sure we're there yet).

Doing it inline would mean that contributors, future, team members, and our future selves would have to remember to bust the cache inline before redirects, only in circumstances. That felt more challenging than recognizing a new, documented pattern of avoiding redirects in a few places. Doing the cache bust inline is a larger code change. We would have to manually call the cache bust, do larger code updates to avoid all the effort of double busting (maybe?), make some performance improvements, and account for errors.

> Do we actually need the cache in those endpoints? Would it that bad if the page the user lands on is not cached? Obviously case by case.
That's tough, that would mean we stop caching pages like comments, articles, listings, etc. which I don't think we want to do.

> Is there anything we can cache server-side and stop using edge in there?
Same thing here. That's a much larger architecture change to try and get that right. Ultimately (AFAIK), we want to remain with edge caching.

I'm certainly not opposed to a larger fix. My thinking was since this is an edge case, caching with Openresty is a work-in-progress, and we're still learning what edge caching looks like with regards to Forem, that it would be better to hold off on a large change to see how things shake out and gather some more intel first.

The core of this issue is we're trying to treat heavily cached content as near-real-time (show updates "live"). Naturally, those (somewhat) clashing mantras lead to what may feel like some strange code.

> adding a way to bypass the cache after an update

With a redirect that's not possible 😢 . I think what we'll have to do is dive pretty deep at the edge to get some logic in place that would understand how to handle these edge cases for us.

Another avenue I considered is where we're redirecting to. The challenge is we're updating cached content in an `update` action and then redirecting to a `show` or `index` action that displays the same cached content. If we redirected to something not cached, or something more generic (like a success page), or back to edit with a success message (a page that isn't cached) then we could keep the Rails idiom _and_ fix this issue. It seemed like where we were redirecting to though was a nice UX so I didn't want to compromise the UX for the sake of sticking to the Rails way in a handful of spots.

**Eventually, we can get some edge sorcery in place to try and do this for us.**

**Comment by atsmith813** [comment]
To be completely honest, I don't think we lose this race condition that often as it is right now. We definitely do lose at times though, it's absolutely an issue.

In other words, I have a hard time re-creating the issue in production because it happens infrequently.

**Comment by citizen428** [comment]
@atsmith813 Thanks for taking the time to add all this context, this makes a lot more sense to me now. 

From my somewhat uninformed vantage point it seemed like documenting the circumstances under which to bust inline is about the same level of complexity for contributors as documenting the circumstances under which not to redirect, but your explanation made me rethink this a bit, especially since you're way more involved with our caching efforts.

**Comment by atsmith813** [comment]
These issues are no longer valid, closing this issue in favor of the new RFC process.
